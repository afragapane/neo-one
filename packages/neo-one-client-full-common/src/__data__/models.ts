import { common, ContractParameterTypeModel, ECPoint, SignatureString, UInt160 } from '@neo-one/client-common';
import { constants } from '@neo-one/utils';
import {
  ContractABIModel,
  ContractEventModel,
  ContractGroupModel,
  ContractManifestModel,
  ContractMethodDescriptorModel,
  ContractModel,
  ContractParameterDeclarationModel,
  ContractPermissionDescriptorModel,
  ContractPermissionsModel,
  ContractPropertyStateModel,
} from '../models';

export const contractParamModel = {
  boolean: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Boolean,
    name: 'param',
  }),
  byteArray: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.ByteArray,
    name: 'param',
  }),
  hash160: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Hash160,
    name: 'param',
  }),
  hash256: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Hash256,
    name: 'param',
  }),
  array: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Array,
    name: 'param',
  }),
  integer: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Integer,
    name: 'param',
  }),
  interopInterface: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.InteropInterface,
    name: 'param',
  }),
  map: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Map,
    name: 'param',
  }),
  publicKey: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.PublicKey,
    name: 'param',
  }),
  signature: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Signature,
    name: 'param',
  }),
  string: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.String,
    name: 'param',
  }),
  void: new ContractParameterDeclarationModel({
    type: ContractParameterTypeModel.Void,
    name: 'param',
  }),
};

export const contractMethodDescriptorModel = (
  parameters: readonly ContractParameterDeclarationModel[] = [contractParamModel.boolean],
  returnType: ContractParameterTypeModel = ContractParameterTypeModel.Void,
  name = 'function',
) => new ContractMethodDescriptorModel({ name, parameters, returnType });

export const contractEventModel = (
  parameters: readonly ContractParameterDeclarationModel[] = [contractParamModel.boolean],
  name = 'event',
) => new ContractEventModel({ name, parameters });

export const contractAbiModel = (
  methods: readonly ContractMethodDescriptorModel[] = [contractMethodDescriptorModel()],
  events: readonly ContractEventModel[] = [contractEventModel()],
  entryPoint: ContractMethodDescriptorModel = contractMethodDescriptorModel(),
  hash = common.bufferToUInt160(Buffer.alloc(20, 1)),
) => new ContractABIModel({ hash, entryPoint, methods, events });

export const contractGroupModel = (
  publicKey: ECPoint = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY),
  signature: SignatureString = 'ccaab040cc25021c91567b75db4778853441869157b8f6aad960cdcf1069812480027a528ca9b98e2205027de20696f848cf81824eeb7af1d5110870870ceb67',
) => new ContractGroupModel({ publicKey, signature });

export const contractPermissionDescriptorModel = (hashOrGroupType: 'uint160' | 'ecpoint' | undefined) => {
  let hashOrGroup: UInt160 | ECPoint | undefined;
  if (hashOrGroupType === 'uint160') {
    hashOrGroup = common.bufferToUInt160(Buffer.alloc(20, 1));
  } else if (hashOrGroupType === 'ecpoint') {
    hashOrGroup = common.stringToECPoint(constants.PRIVATE_NET_PUBLIC_KEY);
  }

  return new ContractPermissionDescriptorModel({ hashOrGroup });
};

export const contractPermissionsModel = (
  hashOrGroupType: 'uint160' | 'ecpoint' | undefined,
  methods: readonly string[] = [],
) => new ContractPermissionsModel({ contract: contractPermissionDescriptorModel(hashOrGroupType), methods });

export const contractManifestModel = (
  groups: readonly ContractGroupModel[] = [contractGroupModel()],
  features: ContractPropertyStateModel = ContractPropertyStateModel.HasStoragePayable,
  abi: ContractABIModel = contractAbiModel(),
  permissions: readonly ContractPermissionsModel[] = [contractPermissionsModel('uint160', ['method1'])],
  trusts: readonly UInt160[] = [common.bufferToUInt160(Buffer.alloc(20, 1))],
  safeMethods: readonly string[] = ['method1', 'method2'],
) => new ContractManifestModel({ groups, features, abi, permissions, trusts, safeMethods });

export const contractModel = (
  script: Buffer = Buffer.alloc(25),
  manifest: ContractManifestModel = contractManifestModel(),
) => new ContractModel({ script, manifest });
