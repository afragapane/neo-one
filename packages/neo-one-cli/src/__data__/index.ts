export * from './getClients';
export * from './getContracts';
