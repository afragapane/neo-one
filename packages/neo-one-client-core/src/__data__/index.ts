export * from './keys';
export * from './data';
