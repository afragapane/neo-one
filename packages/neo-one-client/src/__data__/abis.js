/* @flow */
import type {
  ABIEvent,
  ABIFunction,
  ABIParameter,
  ABIReturn,
  ABI,
} from '../types';

export const returns = {
  Signature: { type: 'Signature' },
  Boolean: { type: 'Boolean' },
  Hash160: { type: 'Hash160' },
  Hash256: { type: 'Hash256' },
  ByteArray: { type: 'ByteArray' },
  PublicKey: { type: 'PublicKey' },
  String: { type: 'String' },
  InteropInterface: { type: 'InteropInterface' },
  Void: { type: 'Void' },
  Integer: { type: 'Integer', decimals: 7 },
  Array: { type: 'Array', value: { type: 'String' } },
};

export const parameters = {
  Signature: { name: 'signatureName', type: 'Signature' },
  Boolean: { name: 'booleanName', type: 'Boolean' },
  Integer: { name: 'intName', type: 'Integer', decimals: 7 },
  Hash160: { name: 'hash160Name', type: 'Hash160' },
  Hash256: { name: 'hash256Name', type: 'Hash256' },
  ByteArray: { name: 'byteName', type: 'ByteArray' },
  PublicKey: { name: 'pubkeyName', type: 'PublicKey' },
  String: { name: 'stringName', type: 'String' },
  Array: {
    name: 'arrayName',
    type: 'Array',
    value: {
      name: 'stringName',
      type: 'String',
    },
  },
  InteropInterface: { name: 'interopName', type: 'InteropInterface' },
  Void: { name: 'voidName', type: 'Void' },
};

export const abiFunction = (
  constant: boolean = false,
  params: Array<ABIParameter> = [parameters.String],
  verify: boolean = false,
  name: string = 'funcName',
  returnType: ABIReturn = returns.Void,
): ABIFunction => ({
  name,
  constant,
  parameters: params,
  verify,
  returnType,
});

export const abiEvent = (
  name: string = 'eventName',
  params: Array<ABIParameter> = [parameters.String],
): ABIEvent => ({
  name,
  parameters: params,
});

export const abi = (
  functions: Array<ABIFunction> = [abiFunction()],
  events?: Array<ABIEvent> = [],
): ABI => ({
  functions,
  events,
});
