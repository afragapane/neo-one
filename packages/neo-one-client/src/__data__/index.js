/* @flow */
export { default as keys } from './keys';
export { default as transactions } from './transactions';
