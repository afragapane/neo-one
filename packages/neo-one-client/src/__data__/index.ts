export * from './keys';
export * from './data';
export * from './transactions';
export * from './factory';
