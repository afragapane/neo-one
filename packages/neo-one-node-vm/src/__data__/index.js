/* @flow */
export { default as createBlockchain } from './createBlockchain';
export { default as keys } from './keys';
export * as transactions from './transactions';
export * as testUtils from './utils';
