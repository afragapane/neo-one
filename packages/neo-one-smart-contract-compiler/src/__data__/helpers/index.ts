export * from './compile';
export * from './createContext';
export * from './execute';
export * from './extractors';
export * from './getMonitor';
export * from './startNode';
export * from './transpileAndExecute';
export * from './utils';
