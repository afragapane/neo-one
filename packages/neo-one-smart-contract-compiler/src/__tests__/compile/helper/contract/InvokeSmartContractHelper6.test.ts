import { createPrivateKey, Hash256, privateKeyToAddress } from '@neo-one/client';
import BigNumber from 'bignumber.js';
import { helpers } from '../../../../__data__';

const properties = `
public readonly properties = {
  codeVersion: '1.0',
  author: 'dicarlo2',
  email: 'alex.dicarlo@neotracker.io',
  description: 'The TestSmartContract',
};
`;

describe('InvokeSmartContractHelper', () => {
  test('class that used processedTransactions, send, receive and claim', async () => {
    const node = await helpers.startNode();
    const accountID = node.masterWallet.account.id;
    const contract = await node.addContract(`
      import { Address, Deploy, SmartContract, send, receive, claim } from '@neo-one/smart-contract';

      export class TestSmartContract extends SmartContract {
        ${properties}

        public constructor(public readonly owner = Deploy.senderAddress) {
          super();
        }

        @send
        public send(): boolean {
          return Address.isCaller(this.owner);
        }

        @receive
        public receive(): boolean {
          return Address.isCaller(this.owner);
        }

        @send
        @receive
        public sendReceive(): boolean {
          return Address.isCaller(this.owner);
        }

        @claim
        public claim(): boolean {
          return Address.isCaller(this.owner);
        }
      }
    `);

    await node.executeString(`
      import { Address, SmartContract } from '@neo-one/smart-contract';

      interface Contract {
        deploy(owner: Address): boolean;
      }
      const contract = SmartContract.for<Contract>(Address.from('${contract.address}'));

      assertEqual(contract.deploy(Address.from('${accountID.address}')), true);
      assertEqual(contract.deploy(Address.from('${accountID.address}')), false);
    `);

    const transferResult = await node.client.transfer([
      {
        asset: Hash256.NEO,
        amount: new BigNumber(10),
        to: contract.address,
      },
    ]);
    await transferResult.confirmed();

    let account = await node.client.read(accountID.network).getAccount(contract.address);
    expect(account.balances[Hash256.NEO].toString()).toEqual('10');

    const smartContract = node.client.smartContract({
      networks: {
        [accountID.network]: {
          address: contract.address,
        },
      },
      abi: {
        functions: [
          {
            name: 'refundAssets',
            send: true,
            parameters: [{ name: 'transactionHash', type: 'Hash256' }],
            returnType: { type: 'Boolean' },
          },
          {
            name: 'claim',
            claim: true,
            parameters: [],
            returnType: { type: 'Boolean' },
          },
          {
            name: 'receive',
            receive: true,
            parameters: [],
            returnType: { type: 'Boolean' },
          },
          {
            name: 'send',
            send: true,
            parameters: [],
            returnType: { type: 'Boolean' },
          },
          {
            name: 'sendReceive',
            send: true,
            receive: true,
            parameters: [],
            returnType: { type: 'Boolean' },
          },
        ],
      },
      sourceMaps: Promise.resolve(node.sourceMaps),
    });

    await expect(
      smartContract.refundAssets(transferResult.transaction.hash, {
        sendFrom: [
          {
            asset: Hash256.NEO,
            amount: new BigNumber(9),
            to: accountID.address,
          },
        ],
      }),
    ).rejects.toBeDefined();

    const count = await node.client.read(accountID.network).getBlockCount();
    await Promise.all([
      node.client.read(accountID.network).getBlock(count, { timeoutMS: 2500 }),
      node.developerClient.runConsensusNow(),
    ]);

    account = await node.client.read(accountID.network).getAccount(contract.address);
    expect(account.balances[Hash256.NEO].toString()).toEqual('10');

    const refundResult = await smartContract.refundAssets(transferResult.transaction.hash, {
      sendFrom: [
        {
          asset: Hash256.NEO,
          amount: new BigNumber(10),
          to: accountID.address,
        },
      ],
    });
    const refundReceipt = await refundResult.confirmed();
    expect(refundReceipt.result.state).toEqual('HALT');

    await expect(
      smartContract.refundAssets(transferResult.transaction.hash, {
        sendFrom: [
          {
            asset: Hash256.NEO,
            amount: new BigNumber(10),
            to: accountID.address,
          },
        ],
      }),
    ).rejects.toBeDefined();

    const claimResult = await smartContract.claim({ claimAll: true });
    await claimResult.confirmed();
    expect(claimResult.transaction.outputs.length).toEqual(1);

    const receiveResult = await smartContract.receive({
      sendTo: [
        {
          asset: Hash256.NEO,
          amount: new BigNumber(20),
        },
      ],
    });
    const receiveReceipt = await receiveResult.confirmed();
    expect(receiveReceipt.result.state).toEqual('HALT');

    const sendResult = await smartContract.send({
      sendFrom: [
        {
          asset: Hash256.NEO,
          amount: new BigNumber(10),
          to: accountID.address,
        },
      ],
    });
    const sendReceipt = await sendResult.confirmed();
    expect(sendReceipt.result.state).toEqual('HALT');

    await expect(
      smartContract.refundAssets(sendResult.transaction.hash, {
        sendFrom: [
          {
            asset: Hash256.NEO,
            amount: new BigNumber(10),
            to: accountID.address,
          },
        ],
      }),
    ).rejects.toBeDefined();

    account = await node.client.read(accountID.network).getAccount(contract.address);
    expect(account.balances[Hash256.NEO].toString()).toEqual('10');

    const sendReceiveResult = await smartContract.sendReceive({
      sendTo: [
        {
          asset: Hash256.GAS,
          amount: new BigNumber(20),
        },
      ],
      sendFrom: [
        {
          asset: Hash256.NEO,
          amount: new BigNumber(10),
          to: accountID.address,
        },
      ],
    });
    const sendReceiveReceipt = await sendReceiveResult.confirmed();
    expect(sendReceiveReceipt.result.state).toEqual('HALT');

    await expect(
      smartContract.refundAssets(sendResult.transaction.hash, {
        sendFrom: [
          {
            asset: Hash256.GAS,
            amount: new BigNumber(20),
            to: accountID.address,
          },
        ],
      }),
    ).rejects.toBeDefined();

    account = await node.client.read(accountID.network).getAccount(contract.address);
    expect(account.balances[Hash256.GAS].toString()).toEqual('20');
    expect(account.balances[Hash256.NEO]).toBeUndefined();
  });

  test('realistic classes that uses forward args and forwarded args', async () => {
    const node = await helpers.startNode();
    const token = await node.addContract(`
      import {
        Address,
        constant,
        Contract,
        createEventNotifier,
        Fixed,
        ForwardValue,
        MapStorage,
        SmartContract,
      } from '@neo-one/smart-contract';

      const notifyTransfer = createEventNotifier<Address | undefined, Address | undefined, Fixed<8>>(
        'transfer',
        'from',
        'to',
        'amount',
      );

      interface TokenPayableContract {
        // tslint:disable-next-line readonly-array
        readonly approveReceiveTransfer: (from: Address, amount: Fixed<0>, asset: Address, ...args: ForwardValue[]) => boolean;
      }

      export class Token extends SmartContract {
        public readonly properties = {
          codeVersion: '1.0',
          author: 'dicarlo2',
          email: 'alex.dicarlo@neotracker.io',
          description: 'NEOâ€¢ONE Token',
        };
        public readonly name = 'One';
        public readonly symbol = 'ONE';
        public readonly decimals = 8;
        private readonly balances = MapStorage.for<Address, Fixed<8>>();
        private mutableSupply: Fixed<8> = 0;

        @constant
        public get totalSupply(): Fixed<8> {
          return this.mutableSupply;
        }

        @constant
        public balanceOf(address: Address): Fixed<8> {
          const balance = this.balances.get(address);

          return balance === undefined ? 0 : balance;
        }

        // tslint:disable-next-line readonly-array
        public transfer(from: Address, to: Address, amount: Fixed<8>, ...approveArgs: ForwardValue[]): boolean {
          if (amount < 0) {
            throw new Error(\`Amount must be greater than 0: \${amount}\`);
          }

          const fromBalance = this.balanceOf(from);
          if (fromBalance < amount) {
            return false;
          }

          if (!Address.isCaller(from)) {
            return false;
          }

          const contract = Contract.for(to);
          if (contract !== undefined && !Address.isCaller(to)) {
            const smartContract = SmartContract.for<TokenPayableContract>(to);
            if (!smartContract.approveReceiveTransfer(from, amount, this.address, ...approveArgs)) {
              return false;
            }
          }

          const toBalance = this.balanceOf(to);
          this.balances.set(from, fromBalance - amount);
          this.balances.set(to, toBalance + amount);
          notifyTransfer(from, to, amount);

          return true;
        }

        public issue(to: Address, amount: Fixed<8>): boolean {
          if (amount < 0) {
            throw new Error(\`Amount must be greater than 0: \${amount}\`);
          }

          const toBalance = this.balanceOf(to);
          this.balances.set(to, toBalance + amount);
          notifyTransfer(undefined, to, amount);
          this.mutableSupply += amount;

          return true;
        }
      }
    `);

    const escrow = await node.addContract(`
      import { Address, constant, Fixed, ForwardedValue, MapStorage, SmartContract } from '@neo-one/smart-contract';

      export class Escrow extends SmartContract {
        public readonly properties = {
          codeVersion: '1.0',
          author: 'dicarlo2',
          email: 'alex.dicarlo@neotracker.io',
          description: 'Escrow account',
        };
        private readonly balances = MapStorage.for<[Address, Address, Address], Fixed<8>>();

        @constant
        public balanceOf(from: Address, to: Address, asset: Address): Fixed<8> {
          const balance = this.balances.get([to, asset, from]);

          return balance === undefined ? 0 : balance;
        }

        public approveReceiveTransfer(from: Address, amount: Fixed<8>, asset: Address, to: ForwardedValue<Address>): boolean {
          if (!Address.isCaller(asset)) {
            return false;
          }

          this.setBalance(from, to, asset, this.balanceOf(from, to, asset) + amount);

          return true;
        }

        private setBalance(from: Address, to: Address, asset: Address, amount: Fixed<8>): void {
          this.balances.set([to, asset, from], amount);
        }
      }
    `);

    await node.executeString(`
      import { Address, SmartContract } from '@neo-one/smart-contract';

      interface Contract {
        deploy(): boolean;
      }
      const token = SmartContract.for<Contract>(Address.from('${token.address}'));
      const escrow = SmartContract.for<Contract>(Address.from('${escrow.address}'));

      assertEqual(token.deploy(), true);
      assertEqual(escrow.deploy(), true);
    `);

    const accountID = node.masterWallet.account.id;
    const tokenContract = node.client.smartContract({
      networks: {
        [accountID.network]: {
          address: token.address,
        },
      },
      abi: {
        functions: [
          {
            name: 'balanceOf',
            constant: true,
            parameters: [{ name: 'address', type: 'Address' }],
            returnType: { type: 'Integer', decimals: 8 },
          },
          {
            name: 'transfer',
            parameters: [
              { name: 'from', type: 'Address' },
              { name: 'to', type: 'Address' },
              { name: 'amount', type: 'Integer', decimals: 8 },
              { name: 'approveArgs', type: 'ForwardValue', rest: true },
            ],
            returnType: { type: 'Boolean' },
          },
          {
            name: 'issue',
            parameters: [{ name: 'to', type: 'Address' }, { name: 'amount', type: 'Integer', decimals: 8 }],
            returnType: { type: 'Boolean' },
          },
        ],
        events: [
          {
            name: 'transfer',
            parameters: [
              {
                type: 'Address',
                name: 'from',
                optional: true,
              },
              {
                type: 'Address',
                name: 'to',
                optional: true,
              },
              {
                type: 'Integer',
                name: 'amount',
                decimals: 8,
              },
            ],
          },
        ],
      },
      sourceMaps: Promise.resolve(node.sourceMaps),
    });

    const escrowContract = node.client.smartContract({
      networks: {
        [accountID.network]: {
          address: escrow.address,
        },
      },
      abi: {
        functions: [
          {
            name: 'balanceOf',
            constant: true,
            parameters: [
              { name: 'from', type: 'Address' },
              { name: 'to', type: 'Address' },
              { name: 'asset', type: 'Address' },
            ],
            returnType: { type: 'Integer', decimals: 8 },
          },
          {
            name: 'approveReceiveTransfer',
            parameters: [
              { name: 'from', type: 'Address' },
              { name: 'amount', type: 'Integer', decimals: 8 },
              { name: 'asset', type: 'Address' },
              { name: 'to', type: 'Address', forwardedValue: true },
            ],
            returnType: { type: 'Boolean' },
          },
        ],
      },
      sourceMaps: Promise.resolve(node.sourceMaps),
    });

    let receipt = await tokenContract.issue.confirmed(accountID.address, new BigNumber('20'));
    if (receipt.result.state === 'FAULT') {
      throw new Error(receipt.result.message);
    }
    const toAddress = privateKeyToAddress(createPrivateKey());
    receipt = await tokenContract.transfer.confirmed(
      accountID.address,
      escrow.address,
      new BigNumber('10'),
      ...escrowContract.forwardApproveReceiveTransferArgs(toAddress),
    );
    if (receipt.result.state === 'FAULT') {
      throw new Error(receipt.result.message);
    }
    const result = await escrowContract.balanceOf(accountID.address, toAddress, token.address);
    expect(result.toString()).toEqual('10');
  });

  test('classes that uses forward args, forwarded args, forward return and forwarded return values', async () => {
    const node = await helpers.startNode();
    const forward = await node.addContract(`
      import {
        Address,
        constant,
        ForwardValue,
        SmartContract,
      } from '@neo-one/smart-contract';

      interface TokenPayableContract {
        readonly forwardTo: (...args: ForwardValue[]) => ForwardValue;
        readonly forwardToConstant: (...args: ForwardValue[]) => ForwardValue;
      }

      export class Forward extends SmartContract {
        public readonly properties = {
          codeVersion: '1.0',
          author: 'dicarlo2',
          email: 'alex.dicarlo@neotracker.io',
          description: 'NEOâ€¢ONE Token',
        };

        public forward(address: Address, ...forwardArgs: ForwardValue[]): ForwardValue {
          const smartContract = SmartContract.for<TokenPayableContract>(address);

          return smartContract.forwardTo(...forwardArgs);
        }

        @constant
        public forwardConstant(address: Address, ...forwardArgs: ForwardValue[]): ForwardValue {
          const smartContract = SmartContract.for<TokenPayableContract>(address);

          return smartContract.forwardToConstant(...forwardArgs);
        }
      }
    `);

    const forwarded = await node.addContract(`
      import { constant, Fixed, Integer, ForwardedValue, SmartContract } from '@neo-one/smart-contract';

      export class Forwarded extends SmartContract {
        public readonly properties = {
          codeVersion: '1.0',
          author: 'dicarlo2',
          email: 'alex.dicarlo@neotracker.io',
          description: 'Escrow account',
        };

        public forwardTo(first: ForwardedValue<Fixed<8>>, second: ForwardedValue<Integer>): ForwardedValue<Fixed<8>> {
          return first * second;
        }

        @constant
        public forwardToConstant(first: ForwardedValue<Fixed<8>>, second: ForwardedValue<Integer>): ForwardedValue<Fixed<8>> {
          return first * second;
        }
      }
    `);

    await node.executeString(`
      import { Address, SmartContract } from '@neo-one/smart-contract';

      interface Contract {
        deploy(): boolean;
      }
      const forward = SmartContract.for<Contract>(Address.from('${forward.address}'));
      const forwarded = SmartContract.for<Contract>(Address.from('${forwarded.address}'));

      assertEqual(forward.deploy(), true);
      assertEqual(forwarded.deploy(), true);
    `);

    const accountID = node.masterWallet.account.id;
    const forwardContract = node.client.smartContract({
      networks: {
        [accountID.network]: {
          address: forward.address,
        },
      },
      abi: {
        functions: [
          {
            name: 'forward',
            parameters: [{ name: 'address', type: 'Address' }, { name: 'args', type: 'ForwardValue', rest: true }],
            returnType: { type: 'ForwardValue' },
          },
          {
            name: 'forwardConstant',
            constant: true,
            parameters: [{ name: 'address', type: 'Address' }, { name: 'args', type: 'ForwardValue', rest: true }],
            returnType: { type: 'ForwardValue' },
          },
        ],
      },
      sourceMaps: Promise.resolve(node.sourceMaps),
    });

    const forwardedContract = node.client.smartContract({
      networks: {
        [accountID.network]: {
          address: forwarded.address,
        },
      },
      abi: {
        functions: [
          {
            name: 'forwardTo',
            parameters: [
              { name: 'first', type: 'Integer', decimals: 8, forwardedValue: true },
              { name: 'second', type: 'Integer', decimals: 0, forwardedValue: true },
            ],
            returnType: { type: 'Integer', decimals: 8, forwardedValue: true },
          },
          {
            name: 'forwardToConstant',
            constant: true,
            parameters: [
              { name: 'first', type: 'Integer', decimals: 8, forwardedValue: true },
              { name: 'second', type: 'Integer', decimals: 0, forwardedValue: true },
            ],
            returnType: { type: 'Integer', decimals: 8, forwardedValue: true },
          },
        ],
      },
      sourceMaps: Promise.resolve(node.sourceMaps),
    });

    const receipt = await forwardContract.forward.confirmed(
      forwarded.address,
      ...forwardedContract.forwardForwardToArgs(new BigNumber('20'), new BigNumber('3')),
    );
    if (receipt.result.state === 'FAULT') {
      throw new Error(receipt.result.message);
    }
    expect(forwardedContract.forwardForwardToReturn(receipt).result.value.toString()).toEqual('60');

    const result = await forwardContract.forwardConstant(
      forwarded.address,
      ...forwardedContract.forwardForwardToConstantArgs(new BigNumber('20'), new BigNumber('3')),
    );
    expect(forwardedContract.forwardForwardToConstantReturn(result).toString()).toEqual('60');
  });
});
