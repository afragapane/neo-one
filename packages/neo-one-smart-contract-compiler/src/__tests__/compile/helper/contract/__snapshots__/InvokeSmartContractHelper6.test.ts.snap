// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 1`] = `"1.055"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 2`] = `"1.889"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 3`] = `"1.898"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 4`] = `"1.578"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 5`] = `"3.381"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 6`] = `"0.727"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 7`] = `"3.35"`;

exports[`InvokeSmartContractHelper class that used processedTransactions, send, receive and claim 8`] = `"0.727"`;
